test/extended/prometheus/prometheus.go

```diff
@@ -42,6 +44,225 @@ import (
+			var err error
+
+			alertingRules, err = fetchAlertingRules(oc, url, bearerToken)
+			if err != nil {
+				e2e.Failf("Failed to fetch alerting rules: %v", err)
+			}
+		}
+	})
+
+	g.It("should have a valid severity label", func() {
+		forEachAlertingRule(alertingRules, func(alert alertingRule) sets.String {
+			severityRe := regexp.MustCompile("^critical|warning|info$")
+
+			severity, found := alert.Labels["severity"]
+			if !found {
+				return sets.NewString("has no 'severity' label")
+			}
+
+			if !severityRe.MatchString(severity) {
+				return sets.NewString(
+					fmt.Sprintf("has a 'severity' label value of %q which doesn't match %q",
+						severity, severityRe.String(),
+					),
+				)
+			}
+
+			return nil
+		})
+	})
+
+	g.It("should have description and summary annotations", func() {
+		forEachAlertingRule(alertingRules, func(alert alertingRule) sets.String {
+			if descriptionExceptions.Has(alert.Name) {
+				framework.Logf("Alerting rule %q is known to have missing annotations.")
+				return nil
+			}
+
+			violations := sets.NewString()
+
+			if _, found := alert.Annotations["description"]; !found {
+				// If there's no 'description' annotation, but there is a
+				// 'message' annotation, suggest renaming it.
+				if _, found := alert.Annotations["message"]; found {
+					violations.Insert("has no 'description' annotation, but has a 'message' annotation." +
+						" OpenShift alerts must use 'description' -- consider renaming the annotation")
+				} else {
+					violations.Insert("has no 'description' annotation")
+				}
+			}
+
+			if _, found := alert.Annotations["summary"]; !found {
+				violations.Insert("has no 'summary' annotation")
+			}
+
+			return violations
+		})
+	})
+
+	g.It("should have a runbook_url annotation if the alert is critical", func() {
+		violations := sets.NewString()
+
+		forEachAlertingRule(alertingRules, func(alert alertingRule) sets.String {
+			severity := alert.Labels["severity"]
+			runbook := alert.Annotations["runbook_url"]
+
+			if severity == "critical" && runbook == "" {
+				violations.Insert(
+					fmt.Sprintf("WARNING: Alert %q is critical and has no 'runbook_url' annotation", alert.Name),
+				)
+			} else if runbook != "" {
+				// If there's a 'runbook_url' annotation, make sure it's a
+				// valid URL and that we can fetch the contents.
+				if err := validateURL(runbook, 10*time.Second); err != nil {
+					violations.Insert(
+						fmt.Sprintf("WARNING: Alert %q has an invalid 'runbook_url' annotation: %v",
+							alert.Name, err),
+					)
+				}
+			}
+
+			return nil // Always return nil, because these aren't required yet.
+		})
+
+		// These aren't enforced yet.  They are just reporting the errors by
+		// calling the g.Skip() method.
+		if len(violations) > 0 {
+			g.Skip(
+				fmt.Sprintf("Incompliant rules detected:\n\n%s", strings.Join(violations.List(), "\n")),
+			)
+		}
+	})
+})
+
+// RuleDiscovery has info for all rules
+type ruleDiscovery struct {
```

Comment:
I think we should move the type definitions and helper functions to https://github.com/openshift/origin/tree/master/test/extended/util/prometheus.

Reply:
Done.